ts-node-plugin/index.js

```diff
@@ -0,0 +1,38 @@
+const {transform} = require("../dist");
+
+const ModuleKindCommonJS = 1;
+
+/**
+ * ts-node transpiler plugin
+ */
+function create(createOptions) {
+  const {nodeModuleEmitKind} = createOptions;
+  const {module} = createOptions.service.config.options;
+  return {
+    transpile(input, transpileOptions) {
+      const {fileName} = transpileOptions;
+      const transforms = ["typescript"];
+      if (module === ModuleKindCommonJS || nodeModuleEmitKind === "nodecjs") {
+        transforms.push("imports");
+      }
+      if (fileName.endsWith(".tsx") || fileName.endsWith(".jsx")) {
+        transforms.push("jsx");
+      }
+
+      const {code, sourceMap} = transform(input, {
+        transforms,
+        sourceMapOptions: {compiledFilename: fileName},
+        filePath: fileName,
+        preserveDynamicImport: nodeModuleEmitKind === "nodecjs",
+        injectCreateRequireForImportRequire: nodeModuleEmitKind === "nodeesm",
+        disableESTransforms: true,
+      });
+      return {
+        outputText: code,
+        sourceMapText: JSON.stringify(sourceMap),
```

Comment:
@cspotcode FYI I'm generating the `sourceMap` as a JS object and stringifying it here, and I see that you're later parsing it, modifying it, and stringifying it again, so it's a little wasteful, though I imagine it would be very unlikely to have a measurable impact on performance.

(Sucrase source maps are a bit silly anyway because every line is guaranteed to map to that same line number, but I believe WebStorm requires having some source map in order for debugger breakpoints to work.)

Reply:
I forget, does sucrase map line numbers but not column numbers?  I might be thinking of something else.

I bet we can adjust ts-node's API to avoid the unnecessary parse-and-stringify.  We can allow transpilers the option of returning either a string or an object.

I have wondered why swc and typescript give us the sourcemap as a string, but I assume it's cuz they do some sort of efficient string-concat emit, so it's faster for them not to build an object.  Not sure though.

Reply:
Sucrase source maps only map line numbers right now, which is good enough for most cases. It doesn't try to keep column numbers consistent (e.g. by turning type tokens into spaces), and I think that would be impossible for some transforms, but it could be possible on a best-effort basis in the future. Currently the lack of column information in source maps causes an issue with Jest inline snapshots ( https://github.com/alangpierce/sucrase/issues/667 ), so would be nice to improve, but as I mention in that issue, the fix isn't super obvious.

Not sure about other tools, but FWIW it looks like TypeScript makes an object and stringifies it: https://github.com/microsoft/TypeScript/blob/46a12fdb8ac36db7ace0fdf1929e1b25600ccce7/src/compiler/sourcemap.ts#L50 . For swc, my guess is that maybe it's easier to pass a string than an object between Rust and JS.